[BUG] Need sanity check with warning G38_PROBE_TARGET & //USE_ZMIN_PLUG

### Did you test the latest `bugfix-2.0.x` code?

Yes, and the problem still exists.

### Bug Description

I use a probe for Z homing. (connected on the BLTouch port of the Creality 4.2.2) 
After homing, software endstops are used and physical endstops are ignored, also for the Z-axis.
Since this makes the Z end stop superfluous, I removed it and planning to use the pin for something else later.

If I now try to implement the whole thing in the FW, I run into some errors when compiling.

```
Compiling .pio\build\STM32F103RET6_creality\src\src\module\settings.cpp.o
In file included from Marlin\src\module\../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin\src\module\../inc/../HAL/HAL.h:30,
                 from Marlin\src\module\../inc/MarlinConfig.h:31,
                 from Marlin\src\module\endstops.h:28,
                 from Marlin\src\module\endstops.cpp:27:
Marlin\src\module\endstops.cpp:944:65: error: 'Z_MIN' was not declared in this scope; did you mean 'Y_MIN'?
  944 |         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
      |                                                                 ^
Marlin\src\module\../inc/../HAL/./STM32/../shared/Marduino.h:42:25: note: in definition of macro '_BV'
   42 | #define _BV(b) (1UL << (b))
      |                         ^
Marlin\src\module\endstops.cpp:863:38: note: in expansion of macro 'SBI'
  863 |   #define _ENDSTOP_HIT(AXIS, MINMAX) SBI(hit_state, _ENDSTOP(AXIS, MINMAX))
      |                                      ^~~
Marlin\src\module\endstops.cpp:607:32: note: in expansion of macro '__ENDSTOP'
  607 | #define _ENDSTOP(AXIS, MINMAX) __ENDSTOP(AXIS, MINMAX)
      |                                ^~~~~~~~~
Marlin\src\module\endstops.cpp:863:53: note: in expansion of macro '_ENDSTOP'
  863 |   #define _ENDSTOP_HIT(AXIS, MINMAX) SBI(hit_state, _ENDSTOP(AXIS, MINMAX))
      |                                                     ^~~~~~~~
Marlin\src\module\endstops.cpp:944:52: note: in expansion of macro '_ENDSTOP_HIT'
  944 |         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
      |                                                    ^~~~~~~~~~~~
*** [.pio\build\STM32F103RET6_creality\src\src\module\endstops.cpp.o] Error 1
```

### Bug Timeline

_No response_

### Expected behavior

It should work or give a sanity check with warning.

### Actual behavior

I disabled G38_PROBE_TARGET.

### Steps to Reproduce

compile FW with my [Configurations.zip](https://github.com/MarlinFirmware/Marlin/files/7117279/Configurations.zip) or use stock CFG and adjust following:

Marlin\Configuration.h
```// #define USE_ZMIN_PLUG
// #define Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN
#define USE_PROBE_FOR_Z_HOMING
#define BLTOUCH
#define Z_SAFE_HOMING
```
Marlin\Configuration_adv.h
```
#define G38_PROBE_TARGET
```



### Version of Marlin Firmware

bugfix-2.0.x(2021-09-06)

### Printer model

_No response_

### Electronics

_No response_

### Add-ons

_No response_

### Bed Leveling

_No response_

### Your Slicer

_No response_

### Host Software

_No response_

### Additional information & file uploads

Just reported it for completeness.

I don't need G38_PROBE_TARGET at the moment and have disabled it.

--- a/Marlin/src/inc/Conditionals_post.h
+++ b/Marlin/src/inc/Conditionals_post.h
@@ -341,21 +341,21 @@
   #ifdef MANUAL_U_HOME_POS
     #define U_HOME_POS MANUAL_U_HOME_POS
   #else
-    #define U_HOME_POS (U_HOME_DIR < 0 ? U_MIN_POS : U_MAX_POS)
+    #define U_HOME_POS TERN(U_HOME_TO_MIN, U_MIN_POS, U_MAX_POS)
   #endif
 #endif
 #if HAS_V_AXIS
   #ifdef MANUAL_V_HOME_POS
     #define V_HOME_POS MANUAL_V_HOME_POS
   #else
-    #define V_HOME_POS (V_HOME_DIR < 0 ? V_MIN_POS : V_MAX_POS)
+    #define V_HOME_POS TERN(V_HOME_TO_MIN, V_MIN_POS, V_MAX_POS)
   #endif
 #endif
 #if HAS_W_AXIS
   #ifdef MANUAL_W_HOME_POS
     #define W_HOME_POS MANUAL_W_HOME_POS
   #else
-    #define W_HOME_POS (W_HOME_DIR < 0 ? W_MIN_POS : W_MAX_POS)
+    #define W_HOME_POS TERN(W_HOME_TO_MIN, W_MIN_POS, W_MAX_POS)
   #endif
 #endif
 

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -920,7 +920,7 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
       #endif
-      #if HAS_Z_AXIS && USE_ZMIN_PLUG
+      #if (HAS_Z_AXIS && USE_ZMIN_PLUG)
         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
       #endif
       G38_did_trigger = true;

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -920,7 +920,7 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
       #endif
-      #if BOTH(HAS_Z_AXIS, USE_ZMIN_PLUG)
+      #if BOTH(HAS_Z_AXIS, HAS_Z_MIN)
         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
       #endif
       G38_did_trigger = true;

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -915,14 +915,22 @@ void Endstops::update() {
   #if HAS_G38_PROBE // TODO (DerAndere): Add support for HAS_I_AXIS
     // For G38 moves check the probe's pin for ALL movement
     if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) == TERN1(G38_PROBE_AWAY, (G38_move < 4))) {
-             if (stepper.axis_is_moving(X_AXIS)) { _ENDSTOP_HIT(X, TERN(X_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(X_AXIS); }
+      G38_did_trigger = true;
+      #if HAS_X_AXIS
+        const bool xmoving = stepper.axis_is_moving(X_AXIS);
+      #endif
       #if HAS_Y_AXIS
-        else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
+        const bool ymoving = stepper.axis_is_moving(Y_AXIS);
       #endif
       #if HAS_Z_MIN
-        else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
+        const bool zmoving = stepper.axis_is_moving(Z_AXIS);
       #endif
-      G38_did_trigger = true;
+      TERN_(HAS_X_AXIS, if (xmoving) _ENDSTOP_HIT(X, TERN(X_HOME_TO_MIN, MIN, MAX)));
+      TERN_(HAS_Y_AXIS, if (ymoving) _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)));
+      TERN_(HAS_Z_MIN,  if (zmoving) _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)));
+      TERN_(HAS_X_AXIS, if (xmoving) planner.endstop_triggered(X_AXIS));
+      TERN_(HAS_Y_AXIS, if (ymoving) planner.endstop_triggered(Y_AXIS));
+      TERN_(HAS_Z_MIN,  if (zmoving) planner.endstop_triggered(Z_AXIS));
     }
   #endif
 

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -922,15 +922,18 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         const bool ymoving = stepper.axis_is_moving(Y_AXIS);
       #endif
-      #if HAS_Z_MIN
+      #if HAS_Z_AXIS
         const bool zmoving = stepper.axis_is_moving(Z_AXIS);
       #endif
-      TERN_(HAS_X_AXIS, if (xmoving) _ENDSTOP_HIT(X, TERN(X_HOME_TO_MIN, MIN, MAX)));
-      TERN_(HAS_Y_AXIS, if (ymoving) _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)));
-      TERN_(HAS_Z_MIN,  if (zmoving) _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)));
-      TERN_(HAS_X_AXIS, if (xmoving) planner.endstop_triggered(X_AXIS));
-      TERN_(HAS_Y_AXIS, if (ymoving) planner.endstop_triggered(Y_AXIS));
-      TERN_(HAS_Z_MIN,  if (zmoving) planner.endstop_triggered(Z_AXIS));
+      #if HAS_X_AXIS
+        if (xmoving) { _ENDSTOP_HIT(X, ENDSTOP); planner.endstop_triggered(X_AXIS); }
+      #endif
+      #if HAS_Y_AXIS
+        if (ymoving) { _ENDSTOP_HIT(Y, ENDSTOP); planner.endstop_triggered(Y_AXIS); }
+      #endif
+      #if HAS_Z_AXIS
+        if (zmoving) { _ENDSTOP_HIT(Z, ENDSTOP); planner.endstop_triggered(Z_AXIS); }
+      #endif
     }
   #endif
 

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -916,24 +916,10 @@ void Endstops::update() {
     // For G38 moves check the probe's pin for ALL movement
     if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) == TERN1(G38_PROBE_AWAY, (G38_move < 4))) {
       G38_did_trigger = true;
-      #if HAS_X_AXIS
-        const bool xmoving = stepper.axis_is_moving(X_AXIS);
-      #endif
-      #if HAS_Y_AXIS
-        const bool ymoving = stepper.axis_is_moving(Y_AXIS);
-      #endif
-      #if HAS_Z_AXIS
-        const bool zmoving = stepper.axis_is_moving(Z_AXIS);
-      #endif
-      #if HAS_X_AXIS
-        if (xmoving) { _ENDSTOP_HIT(X, ENDSTOP); planner.endstop_triggered(X_AXIS); }
-      #endif
-      #if HAS_Y_AXIS
-        if (ymoving) { _ENDSTOP_HIT(Y, ENDSTOP); planner.endstop_triggered(Y_AXIS); }
-      #endif
-      #if HAS_Z_AXIS
-        if (zmoving) { _ENDSTOP_HIT(Z, ENDSTOP); planner.endstop_triggered(Z_AXIS); }
-      #endif
+      #define _G38_SET(Q) | (stepper.axis_is_moving(_AXIS(Q)) << _AXIS(Q))
+      #define _G38_RESP(Q) if (moving[_AXIS(Q)]) { _ENDSTOP_HIT(Q, ENDSTOP); planner.endstop_triggered(_AXIS(Q)); }
+      const Flags<NUM_AXES> moving = { value_t(NUM_AXES)(0 MAIN_AXIS_MAP(_G38_SET)) };
+      MAIN_AXIS_MAP(_G38_RESP);
     }
   #endif
 

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -920,7 +920,7 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
       #endif
-      #if HAS_Z_AXIS
+      #if HAS_Z_AXIS && USE_ZMIN_PLUG
         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
       #endif
       G38_did_trigger = true;

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -913,7 +913,7 @@ void Endstops::update() {
   #endif
 
   #if HAS_G38_PROBE // TODO (DerAndere): Add support for HAS_I_AXIS
-    // For G38 moves check the probes pin for ALL movement
+    // For G38 moves check the probe's pin for ALL movement
     if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) == TERN1(G38_PROBE_AWAY, (G38_move < 4))) {
       G38_did_trigger = true;
       #define _G38_SET(Q) | (stepper.axis_is_moving(_AXIS(Q)) << _AXIS(Q))

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -920,7 +920,7 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
       #endif
-      #if (HAS_Z_AXIS && USE_ZMIN_PLUG)
+      #if BOTH(HAS_Z_AXIS, USE_ZMIN_PLUG)
         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
       #endif
       G38_did_trigger = true;

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -920,7 +920,7 @@ void Endstops::update() {
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }
       #endif
-      #if BOTH(HAS_Z_AXIS, HAS_Z_MIN)
+      #if HAS_Z_MIN
         else if (stepper.axis_is_moving(Z_AXIS)) { _ENDSTOP_HIT(Z, TERN(Z_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Z_AXIS); }
       #endif
       G38_did_trigger = true;

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -913,9 +913,8 @@ void Endstops::update() {
   #endif
 
   #if HAS_G38_PROBE // TODO (DerAndere): Add support for HAS_I_AXIS
-    #define _G38_OPEN_STATE TERN(G38_PROBE_AWAY, (G38_move >= 4), LOW)
     // For G38 moves check the probe's pin for ALL movement
-    if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) != _G38_OPEN_STATE) {
+    if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) == TERN1(G38_PROBE_AWAY, (G38_move < 4))) {
              if (stepper.axis_is_moving(X_AXIS)) { _ENDSTOP_HIT(X, TERN(X_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(X_AXIS); }
       #if HAS_Y_AXIS
         else if (stepper.axis_is_moving(Y_AXIS)) { _ENDSTOP_HIT(Y, TERN(Y_HOME_TO_MIN, MIN, MAX)); planner.endstop_triggered(Y_AXIS); }

--- a/Marlin/src/module/endstops.cpp
+++ b/Marlin/src/module/endstops.cpp
@@ -913,7 +913,7 @@ void Endstops::update() {
   #endif
 
   #if HAS_G38_PROBE // TODO (DerAndere): Add support for HAS_I_AXIS
-    // For G38 moves check the probe's pin for ALL movement
+    // For G38 moves check the probes pin for ALL movement
     if (G38_move && TEST_ENDSTOP(_ENDSTOP(Z, TERN(USES_Z_MIN_PROBE_PIN, MIN_PROBE, MIN))) == TERN1(G38_PROBE_AWAY, (G38_move < 4))) {
       G38_did_trigger = true;
       #define _G38_SET(Q) | (stepper.axis_is_moving(_AXIS(Q)) << _AXIS(Q))

